altair_pandas/_misc.py

```diff
@@ -0,0 +1,73 @@
+import altair as alt
+from typing import Iterable, Union, List
+
+toltipList = List[alt.Tooltip]
+
+
+def scatter_matrix(
+    df,
+    color: Union[str, Iterable, None] = None,
+    alpha: float = 1.0,
+    tooltip: Union[List[str], toltipList, None] = None,
+):
+    """ plots a scatter matrix
+
+    At the moment does not support neither histogram nor kde;
+    Uses f-f scatterplots instead. Interactive and with a cusotmizable
+    tooltip
+
+    Parameters
+    ----------
+    df : DataFame
+        DataFame to be used for scatterplot. Only numeric columns will be included.
+    color : string or iterable [optional]
+        Can be a column name, specific color value (hex, webcolors), or an array
+        of values to be used.
+    alpha : float
+        Opacity of the markers, within [0,1]
+    tooltip: list [optional]
+        List of specific column names or alt.Tooltip objects. If none (default),
+        will show all columns.
+    """
+    dfc = df.copy()  # otherwise passing array will be preserved
+    cols = dfc._get_numeric_data().columns.astype(str).tolist()
+    tooltip = tooltip or dfc.columns.astype(str).tolist()
+
+    if color is None:
+        pass
+    elif isinstance(color, str):
+        if color in df.columns.astype(str).tolist():
+            pass
+        else:
+            color = alt.value(color)
+    elif hasattr(color, "__len__") and len(color) == len(df):
+        allcols = set(dfc.columns.astype(str))
+        colname = "__color__"
+
+        while True:
+            if colname in allcols:
+                colname += "_"
+            else:
+                break
+
+        dfc[colname] = color
+        color = colname
+    else:
+        raise ValueError(color)
+
+    chart = (
+        alt.Chart(dfc)
+        .mark_circle()
+        .encode(
+            x=alt.X(alt.repeat("column"), type="quantitative"),
+            y=alt.X(alt.repeat("row"), type="quantitative"),
+            opacity=alt.value(alpha),
+            tooltip=tooltip,
+        )
+        .properties(width=150, height=150)
+    )
+
+    if color:
+        chart = chart.encode(color=color)
+
+    return chart.repeat(row=cols, column=cols[::-1]).interactive()
```

Comment:
`chart.repeat(row=cols, column=cols)` can be same I suppose? [matplotlib](https://pandas.pydata.org/pandas-docs/stable/user_guide/visualization.html#scatter-matrix-plot) has them on other diagonal.

Reply:
There is an Altair bug (?) that prevents from doing that :-)
Note that the same is done in Altair example:

https://altair-viz.github.io/gallery/scatter_matrix.html

Reply:
Ah, didn't know about that. `chart.repeat(row=cols, column=cols)` seems to work for me, on my versions..

Reply:
I don't think there's any bug - it just decides the orientation of the diagonals

Reply:
Hm. here is my Issue (am I missing something?):

```
df = pd.DataFrame(np.random.randn(100,3).cumsum(0), columns=['x','y','z'])
cols = df.columns.tolist()


alt.Chart(df).mark_circle().encode(
    alt.X(alt.repeat("column"), type='quantitative'),
    alt.Y(alt.repeat("row"), type='quantitative'),
    color='Origin:N'
).properties(
    width=150,
    height=150
).repeat(
    row=cols,
    column=cols
).interactive()
```

This results in this error:
```
Javascript Error: Expression parse error: {"x": selector003_x, "x": selector003_x_1, "z": selector003_z}
```
Altair 2.3.0

Reply:
What frontend renderer are you using?

Reply:
renderer jupyterlab
jupyterlab                1.0.2